webui/react/src/stores/experiments.tsx

```diff
@@ -1,63 +1,110 @@
 import { Map } from 'immutable';
-import React, { createContext, PropsWithChildren, useCallback, useContext, useState } from 'react';
 
 import { getExperiments } from 'services/api';
 import { V1Pagination } from 'services/api-ts-sdk';
 import { GetExperimentsParams } from 'services/types';
-import { ExperimentItem } from 'types';
+import { ExperimentItem, ExperimentPagination } from 'types';
 import handleError from 'utils/error';
 import { Loadable, Loaded, NotLoaded } from 'utils/loadable';
+import { Observable, observable, WritableObservable } from 'utils/observable';
 import { encodeParams } from 'utils/store';
 
-type ExperimentPagination = {
-  experiments: ExperimentItem[];
-  pagination?: V1Pagination;
+type ExperimentsCache = {
+  experimentIds: Readonly<number[]>;
+  pagination: Readonly<V1Pagination>;
 };
 
-type ExperimentsContext = {
-  experimentsCache: Map<string, ExperimentPagination>;
-  updateExperimentsCache: (ws: Map<string, ExperimentPagination>) => void;
-};
-
-const ExperimentsContext = createContext<ExperimentsContext | null>(null);
-
-export const ExperimentsProvider: React.FC<PropsWithChildren> = ({ children }) => {
-  const [experimentsCache, updateExperimentsCache] = useState<Map<string, ExperimentPagination>>(
-    Map<string, ExperimentPagination>(),
-  );
-  return (
-    <ExperimentsContext.Provider value={{ experimentsCache, updateExperimentsCache }}>
-      {children}
-    </ExperimentsContext.Provider>
-  );
-};
+// This is singleton class
+export class ExperimentsService {
+  static #isInternalConstructing = false;
+  static #instance: ExperimentsService;
+  #experimentsCache: WritableObservable<Map<string, ExperimentsCache>> = observable(Map());
+  #experimentMap: WritableObservable<Map<number, ExperimentItem>> = observable(Map());
 
-export const useFetchExperiments = (
-  params: GetExperimentsParams,
-  canceler: AbortController,
-): (() => Promise<void>) => {
-  const context = useContext(ExperimentsContext);
-  if (context === null) {
-    throw new Error('Attempted to use useFetchExperiments outside of Experiment Context');
+  private constructor() {
+    if (!ExperimentsService.#isInternalConstructing) {
+      throw new TypeError('ExperimentsService is not constructable');
+    }
   }
-  const { experimentsCache, updateExperimentsCache } = context;
-
-  return useCallback(async (): Promise<void> => {
-    try {
-      const response = await getExperiments(params, { signal: canceler.signal });
-      updateExperimentsCache(experimentsCache.set(encodeParams(params), response));
-    } catch (e) {
-      handleError(e);
+
+  // Get singleton instance
+  public static getInstance(): ExperimentsService {
```

Comment:
What's the reason for this vs just `const experimentStore = new ExperimentsService()` at the top-level?

Reply:
one of reasons is to try a different pattern instead of using context. 
other reason is that we can call the class anywhere, but  singleton pattern has some down sides. im open to use context

Reply:
This is still a singleton. What I'm asking is why this static method instead of just
```typescript
const experimentStore = new ExperimentsService()

export default experimentStore;
```

Reply:
i see.  i was thinking about doing it, but this cannot prevent human error if someone wanna export the class  just in case, and i believe factory method is common for singleton class. I just followed the convention.
Mozilla also has the example.
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes/Private_class_fields#simulating_private_constructors

Reply:
If you really want to prevwnt people from constructing it you can do the same thing with a top-level boolean. That said, I don't think you should worry about that, this isn't a library and if somebody does something wrong we can correct that in the PR.

Reply:
fixed